misk/src/main/kotlin/misk/hibernate/ReflectionQueryFactory.kt

```diff
@@ -144,79 +120,75 @@ internal class ReflectionQuery<T : DbEntity<T>>(
     }
   }
 
-  /** Queries for a data class. */
-  class ProjectionHandler<P : Projection>(
-    val constructor: KFunction<P>,
-    val properties: List<List<String>>
-  ) {
-    fun select(
-      criteriaBuilder: CriteriaBuilder,
-      queryRoot: Root<*>
-    ) = criteriaBuilder.array(*properties.map { queryRoot.traverse<Any?>(it) }.toTypedArray())
-
-    fun toValue(row: Any): P {
-      return if (properties.size == 1) {
-        constructor.call(row)
-      } else {
-        constructor.call(*(row as Array<*>))
-      }
-    }
+  class SelectMethodHandler(
+    private val returnList: Boolean,
+    private val constructor: KFunction<*>?,
+    private val properties: List<List<String>>
+  ) : QueryMethodHandler {
 
-    companion object {
-      fun <P : Projection> create(projectionClass: KClass<P>): ProjectionHandler<P> {
-        val errors = mutableListOf<String>()
-        val constructor = projectionClass.primaryConstructor
+    override fun invoke(reflectionQuery: ReflectionQuery<*>, args: Array<out Any>): Any? {
+      val session = args[0] as Session
 
-        val parameters = if (constructor != null) {
-          constructor.parameters
-        } else {
-          errors.add("this type has no primary constructor")
-          listOf()
-        }
+      val criteriaBuilder = session.hibernateSession.criteriaBuilder
+      val query = criteriaBuilder.createQuery(Any::class.java)
+      val root: Root<*> = query.from(reflectionQuery.rootEntityType.java)
 
-        val properties = mutableListOf<List<String>>()
-        for (parameter in parameters) {
-          val property = parameter.findAnnotation<Property>()
-          if (property == null) {
-            errors.add("parameter ${parameter.index} is missing a @Property annotation")
-            continue
-          }
+      query.select(select(criteriaBuilder, root))
+      query.where(reflectionQuery.buildWherePredicate(root, criteriaBuilder))
+      val rows = session.hibernateSession.createQuery(query).list()
+      val list = rows.map { toValue(it) }
 
-          if (!property.value.matches(PATH_PATTERN)) {
-            errors.add("parameter ${parameter.index} path is not valid: '${property.value}'")
-            continue
-          }
-          val path = property.value.split('.')
+      if (returnList) return list
 
-          properties.add(path)
-        }
+      require(list.size <= 1) { "expected at most 1 result but was $list" }
+      return list.firstOrNull()
+    }
 
-        require(errors.isEmpty()) {
-          "Projection class ${projectionClass.java.name} has problems:" +
-              "\n  ${errors.joinToString(separator = "\n  ")}"
-        }
+    private fun select(
+      criteriaBuilder: CriteriaBuilder,
+      queryRoot: Root<*>
+    ) = criteriaBuilder.array(*properties.map { queryRoot.traverse<Any?>(it) }.toTypedArray())
 
-        return ProjectionHandler(constructor!!, properties)
+    private fun toValue(row: Any): Any? {
+      return when {
```

Comment:
for custom types (eg. `IdType`) will the row here already be the correct type due to Hibernate?

Reply:
yep! Yay!